docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,249 @@
+
+Before using a table descriptor for a DML operation (i.e. `SELECT`,
+`INSERT`, `UPDATE`, `DELETE`, etc), the operation needs to obtain a
+read lease for the descriptor. The lease will be guaranteed valid for
+a significant period of time (on the order of minutes) and the lease
+period can be extended as long as a scheme change is not
+performed. When the operation completes it will release the read
+lease.
+
+# Detailed design
+
+Table descriptors will be extended with a version number that is
+incremented on every change to the descriptor:
+
+```proto
+message TableDescriptor {
+  ...
+  optional uint32 id;
+  optional uint32 version;
+  ...
+}
+```
+
+Leases will be tied to a specific version of the descriptor. Lease
+state will be stored in a new `system.lease` table:
+
+```sql
+CREATE TABLE system.lease (
+  DescID     INT,
+  Version    INT,
+  NodeID     INT,
+  Expiration TIMESTAMP,
+  PRIMARY KEY (DescID, Version, NodeID, Expiration)
+)
+```
+
+Entries in the lease table will be added and removed as leases are
+acquired and released. A background goroutine will periodically delete
+expired leases. 
+
+Leases will be granted for a duration measured in minutes (we'll
+assume 5m for the rest of this doc, though experimentation may tune
+this number). A node will acquire a lease before using it in an
+operation and may release the lease when the last local operation
+completes that was using the lease.
+
+The leader of the range containing a table descriptor will gossip the
+most recent version of that table descriptor using the gossip key
+`"table-<descID>"` and the value will be the version number. The
+gossiping of the most recent table versions allows nodes to
+asynchronously discover when a new table version is written. But note
+that it is not necessary for correctness as the protocol for acquiring
+a lease ensures that only the most recent version of a descriptor can
+have a new lease granted on it.
+
+Lease acquisition will perform the following steps in a transaction:
+
+* `SELECT descriptor FROM system.descriptor WHERE id = <descID>`
+* `INSERT INTO system.lease VALUES (<desc.ID>, <desc.version>, <nodeID>, <expiration>)`
+
+Nodes will maintain a map from `<descID, version>` to
+`<TableDescriptor, expiration, localRefCount>`. The local reference
+count will be incremented when a transaction first uses a table and
+decremented when the transaction commits/aborts. When the node
+discovers a new version of the table, either via gossip or by
+acquiring a lease and discovering the version has changed it can
+release the lease on the old version when there are no more local
+references:
+
+* `DELETE FROM system.lease WHERE (DescID, Version, NodeID, Expiration) = (<descID>, <version>, <nodeID>, <expiration>)`
+
+A schema change operation needs to ensure that there is only one
+version of a descriptor in use in the cluster before a mutation of the
+descriptor. The operation will perform the following steps
+transactionally:
```

Comment:
> Yes, we could relax this, but like the F1 paper states: the extra complexity does not seem worth it.

That's not where I was going, I was wondering that since a large schema change consists of multiple steps, but we only lock inbetween steps, that there's no reason that multiple schema changes at the time couldn't start racing each other (so you end up with a random schema, and two confused coordinators). Also, what happens with a schema change coordinator that dies in the middle of a change? That's going to leave the table in an awkward position, and one straightforward way to restore order would be continuing the partial schema change if it had been fully laid out before. Basically I'm suggesting an option to lay out multiple future versions in one go. I don't really care about this feature as much as about handling things going wrong.
If nothing else, we'll probably need to be able to mark schema changes as "intermediate" or "final", so that a stuck intermediate schema change can be rolled back (i.e. another change added which undoes the prior).